### PR TITLE
fix flower and foliage generation

### LINKED ISSUE #238: Add the Sunflower Plains Biome
Exactly the same as Plains, but can spawn Sunflowers

### LINKED ISSUE #241: Add the Flower Forest Biome
Forest variant with several of the new flowers, including peonies, orange tulips, white tulips, oxeye daisies, rose bush and allium

### PR DESCRIPTION
**Pull request by mBornand**
Activate flower generation.
Fixes #238
Fixes #241

The main  base of the code was already present, but it contained some mistakes.

- [x] Flower generation works with default settings.
- [x] Add tulip generation
- [x] stop incorrect mushroom generation

### CODE REVIEW

File: src/Generating/ComposableGenerator.cpp [hunk 1]
```diff
@@ -226,7 +226,8 @@ void cComposableGenerator::InitializeGeneratorDefaults(cIniFile & a_IniFile, eDi
 				"DeadBushes, "
 				"NaturalPatches, "
 				"PreSimulator, "
-				"Animals"
+				"Animals, "
```
Comment (peterbell10) [inline src/Generating/ComposableGenerator.cpp:229]:
```suggestion
				"Animals",
```

File: src/Generating/FinishGen.cpp [hunk 2]
```diff
@@ -818,22 +819,11 @@ void cFinishGenSprinkleFoliage::GenFinish(cChunkDesc & a_ChunkDesc)
...
-					}
-					else if (val1 + val4 > 0.2f)
-					{
-						a_ChunkDesc.SetBlockType(x, ++Top, z, E_BLOCK_BROWN_MUSHROOM);
-					}
+					if ((val1 + val2 > 0.2f) ||
+						(val2 + val3 > 0.2f) ||
+						(val3 + val4 > 0.2f) ||
+						(val1 + val4 > 0.2f))
+					{}
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:826]: This construction looks confusing, is it to remove the hardcoded block values?

Comment (mBornand) [reply src/Generating/FinishGen.cpp:826]: Yes, and to not alter the generation of the other blocks that are generated in the same function.

Comment (mBornand) [reply src/Generating/FinishGen.cpp]: I have tried it without this code. It seems to work as well. It is in the new commit.
Tall grass was also generated twice.

**Comment by tigerw** [comment]
Looking

File: src/Generating/FinishGen.cpp [hunk 3]
```diff
@@ -252,20 +252,20 @@ void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int
 		int z = a_CenterZ + (((rnd / 256) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
 		int Top = a_ChunkDesc.GetHeight(x, z);
 
-		if (a_ChunkDesc.GetBlockType(x, Top, z) != E_BLOCK_GRASS)
+		auto GroundBlockType = a_ChunkDesc.GetBlockType(x, Top, z);
+		if ((GroundBlockType == E_BLOCK_GRASS) ||
+			((GroundBlockType == E_BLOCK_MYCELIUM) && ((a_BlockType == E_BLOCK_RED_MUSHROOM) || (a_BlockType == E_BLOCK_BROWN_MUSHROOM))))
 		{
-			continue;
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:257]: The condition should be formatted
```cpp
if (
    (GroundBlockType == GRASS) ||
    ((X) && (Y || Z))
)
{
...
}

File: src/Generating/FinishGen.cpp [hunk 4]
```diff
@@ -252,20 +252,20 @@ void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int
...
-			a_ChunkDesc.SetBlockTypeMeta(x, Top + 2, z, E_BLOCK_BIG_FLOWER, E_META_BIG_FLOWER_TOP);
-			a_ChunkDesc.SetHeight(x, z, static_cast<HEIGHTTYPE>(Top + 2));
-		}
-		else
-		{
-			a_ChunkDesc.SetHeight(x, z, static_cast<HEIGHTTYPE>(Top + 1));
+			a_ChunkDesc.SetBlockTypeMeta(x, Top + 1, z, a_BlockType, a_BlockMeta);
+			if (a_IsDoubleTall)
+			{
+				a_ChunkDesc.SetBlockTypeMeta(x, Top + 2, z, E_BLOCK_BIG_FLOWER, E_META_BIG_FLOWER_TOP);
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:262]: Just to confirm, mycelium mushrooms are never a_IsDoubleTall, right?

**Comment by tigerw** [comment]
It seems cFinishGenSprinkleFoliage has been somewhat gutted, was all of its functionality preserved in the migration to FinishGenClumpTopBlock? E.g. there is a `// Nothing grows above Y=250` on line 863 of FinishGen.cpp, does the new code have/need this check, etc.

**Comment by bearbin** [comment]
@mBornand Any chance that you are going to get back to this, I would like to get this merged in.

File: src/Generating/FinishGen.cpp (hunk 4 above)
Comment (12xx12) [reply src/Generating/FinishGen.cpp:262]: Yes the code checks when calling ```TryPlaceFoliageClump``` if the block is a Tall flower: 
```TryPlaceFoliageClump(a_ChunkDesc, PosX, PosZ, Block.m_BlockType, Block.m_BlockMeta, Block.m_BlockType == E_BLOCK_BIG_FLOWER);```

**Comment by 12xx12** [comment]
> It seems cFinishGenSprinkleFoliage has been somewhat gutted, was all of its functionality preserved in the migration to FinishGenClumpTopBlock? E.g. there is a `// Nothing grows above Y=250` on line 863 of FinishGen.cpp, does the new code have/need this check, etc.

There weren't any checksat all. If the terrain would have caused an invalid height for the placed block the server would have crashed

File: src/Generating/FinishGen.h [hunk 5]
```diff
@@ -181,6 +181,9 @@ class cFinishGenTallGrass :
 	// cFinishGen override:
 	virtual void GenFinish(cChunkDesc & a_ChunkDesc) override;
 
+	bool CanFernGrow(EMCSBiome a_Biome);
+	bool CanLargeFernGrow(EMCSBiome a_Biome);
+
 	int GetBiomeDensity(EMCSBiome a_Biome)
```
Comment (12xx12) [inline src/Generating/FinishGen.h:187]: make this static
```suggestion
	static bool CanFernGrow(EMCSBiome a_Biome);
	static bool CanLargeFernGrow(EMCSBiome a_Biome);

	static int GetBiomeDensity(EMCSBiome a_Biome)
```

**Review by 12xx12 — CHANGES_REQUESTED** [review]

**Review by 12xx12 — APPROVED** [review]

File: src/Generating/FinishGen.cpp [hunk 6]
```diff
@@ -278,7 +287,7 @@ void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int
 void cFinishGenClumpTopBlock::ParseConfigurationString(const AString & a_RawClumpInfo, std::vector<BiomeInfo> & a_Output)
 {
 	// Initialize the vector for all biomes.
-	for (int i = static_cast<int>(a_Output.size()); i < static_cast<int>(biMaxVariantBiome); i++)
+	for (int i = static_cast<int>(a_Output.size()); i <= static_cast<int>(biMaxVariantBiome); i++)
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:290]: what this by the way?

File: src/Generating/FinishGen.cpp [hunk 7]
```diff
@@ -247,25 +247,34 @@ void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int
...
-		int x = a_CenterX + (((rnd % 256) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
-		int z = a_CenterZ + (((rnd / 256) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
+		int Rnd = m_Noise.IntNoise2DInt(ChunkX + ChunkZ + i, ChunkX - ChunkZ - i) / 59;
+		int x = a_CenterX + (((Rnd % 59) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
+		int z = a_CenterZ + (((Rnd / 97) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
 		int Top = a_ChunkDesc.GetHeight(x, z);
 
-		if (a_ChunkDesc.GetBlockType(x, Top, z) != E_BLOCK_GRASS)
+		// Doesn't place if the blocks can't be placed. Checked value also depends on a_IsDoubleTall
+		if (!cChunkDef::IsValidHeight(Top + 1 + static_cast<int>(a_IsDoubleTall)))
```
Comment (tigerw) [inline src/Generating/FinishGen.cpp:256]: `(Top + 1 + (a_IsDoubleTall ? 1 : 0)) >= cChunkDef::Height` may be slightly clearer

File: src/Generating/FinishGen.h [hunk 8]
```diff
@@ -132,7 +132,7 @@ class cFinishGenClumpTopBlock :
 	/** The maximum number of foliage per clump */
 	const int MAX_NUM_FOLIAGE = 8;
 
-	/** The mininum number of foliage per clump */
+	/** The minimnum number of foliage per clump */
```
Comment (tigerw) [inline src/Generating/FinishGen.h:135]: still a typo, minimum

**Comment by tigerw** [comment]
yeah seems good, not sure if anyone has any screenshots for the record?

**Comment by 12xx12** [comment]
i'm gonna make some

File: src/Generating/FinishGen.cpp (hunk 6 above)
Comment (12xx12) [reply src/Generating/FinishGen.cpp:290]: when using biMesaPlateauM the server crashed

**Comment by 12xx12** [comment]
This PR:

Current Master:

**Review by tigerw — APPROVED** [review]

**Pull request merged**